benchmark/benchmark.py

```diff
@@ -0,0 +1,534 @@
+        fn = parallel_bert.main
+
+    sys.stdout.write(f"++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
+    sys.stdout.write(f"BENCHMARK RUNNING ON {device_name}... parallelism {parallel} with batch size {batch_size}...\n")
+    sys.stdout.flush()
+
+    # Spawn a new python interpreter to ensure a nice release of resources
+    mp = pmp.get_context('spawn')
+    rq = mp.SimpleQueue()
+    process = mp.Process(target=wrap, args=(fn, args, rq), daemon=False)
+    process.start()
+    process.join()
+    if process.exitcode != 0:
+        raise Exception(f"Train process exited with exitcode {process.exitcode}")
+    r = rq.get()
+
+    #if model is not Models.none and parallel is Parallelism.pipeline:
+    #    torch.distributed.rpc.shutdown()
+
+    return RunData(r.start_time, r.end_time, model, parallel, epochs, batch_size, r.loss, r.accuracy, device_count, device_model)
+
+
+def uniq(data: typing.List[RunData]):
+    data2 = []
+    idx = 0
+    # parse every element in the list (unless those removed during the process)
+    while idx < len(data):
+        jdx = idx + 1
+        keep = data[idx] # set current data as kept
+        # parse every further element in the list (unless those removed during the process)
+        while jdx < len(data):
+            # if the currently kept element and the current element are of the same "class" ...
+            if data[jdx].same(keep):
+                # ... compare if the current element is greater than the kept one ...
+                if data[jdx].gt(keep):
+                    # ... and keep and remove the current element if it is greater
+                    keep = data.pop(jdx)
+                else:
+                    # ... or only remove the current element if it is not greater
+                    del data[jdx]
+            else:
+                jdx += 1
+        data2.append(keep)
+        idx += 1
+    return data2
+
+
+def plot_training(run_data: typing.List[RunData], imgfile: pathlib.Path, dev_name: str):
+    plt_data = {}
+    for d in run_data:
+        if d.parallel not in plt_data:
+            plt_data[d.parallel] = {}
+        plt_data[d.parallel][d.batch_size] = d.rt_mms()
+
+    fig = plt.figure(figsize=(PLOT_IMG_WIDTH, PLOT_IMG_HEIGHT))
+    ax = fig.add_subplot(1, 1, 1)
+    plots = []
+    legends = []
+    for parallel, v in plt_data.items():
+        p, = ax.plot(v.keys(), v.values(), linestyle=PLOT_LINES[0], linewidth=1.5, color=parallel.colour(), marker=PLOT_MARKERS[0])
+        plots.append(p)
+        legends.append(parallel.to_text())
+
+    ax.legend(plots, legends, loc="upper right")
+    ax.set_title(f"{d.model} -- Run time per batch size -- {dev_name}")
+    ax.set_xlabel("Batch size")
+    ax.set_ylabel("Run time [s]")
+    fig.subplots_adjust(left=PLOT_IMG_MARGIN_LEFT)
+    plt.savefig(str(imgfile))
+
+
+def plot_csv(run_data: typing.List[RunData], img_dir: pathlib.Path, now: datetime.datetime):
```

Comment:
This function has a lot of repeated code. Could you consider refactoring to reduce duplication and complexity?

Reply:
I still think that **four nested loops** are too complicated and should be refactored.